src/main/java/org/openjdk/jextract/impl/HeaderFileBuilder.java

```diff
@@ -393,22 +394,34 @@ class Holder {
 
     private void emitConstant(Class<?> javaType, String constantName, Object value, Declaration declaration) {
         incrAlign();
-        appendLines(STR."""
-            private static final \{javaType.getSimpleName()} \{constantName} = \{constantValue(javaType, value)};
-
-            """);
-        emitDocComment(declaration);
-        appendLines(STR."""
-            public static \{javaType.getSimpleName()} \{constantName}() {
-                return \{constantName};
-            }
-            """);
+        boolean useHolderClass = value instanceof String;
+        if (useHolderClass) {
+            emitDocComment(declaration);
+            appendLines(STR."""
+                public static \{javaType.getSimpleName()} \{constantName}() {
+                    class Holder {
+                        static final \{javaType.getSimpleName()} \{constantName} = \{constantValue(javaType, value)};
+                    }
+                    return Holder.\{constantName};
+                }
+                """);
+        } else {
+            appendLines(STR."""
+                private static final \{javaType.getSimpleName()} \{constantName} = \{constantValue(javaType, value)};
+                """);
+            emitDocComment(declaration);
+            appendLines(STR."""
+                public static \{javaType.getSimpleName()} \{constantName}() {
+                    return \{constantName};
+                }
+                """);
+        }
         decrAlign();
     }
 
     private String constantValue(Class<?> type, Object value) {
         if (value instanceof String) {
-            return STR."Arena.ofAuto().allocateFrom(\"\{Utils.quote(Objects.toString(value))}\")";
+            return STR."\{runtimeHelperName()}.LIBRARY_ARENA.allocateFrom(\"\{Utils.quote(Objects.toString(value))}\")";
```

Comment:
I think I'd prefer to inline this in the above method `emitConstant` - as we're not really supposed to call allocation methods from plain field initializers (e.g. this is only for holder class).